Added different download options and url import/exports

rdiaz738 opened the pull request:
Added the option to configure downloads as either sequential, limited or concurrent. Also added import, export, or copy buttons so you can add a list of urls if you have them. 
**NOTE!!!!** There is a bug where if you select all the items that are downloading and queued, and click cancel selected the next queued item will still download. I think it's because as a downloaded item is canceled the status is being changed for the next queued item and it ends up being skipped. If that's solved then it should be good

**Sequential:**
![2025-03-01 18_05_38-Window](https://github.com/user-attachments/assets/d1450239-1416-477a-b90a-151eb49a87d5)

**Concurrent**
![2025-03-01 18_04_21-Window](https://github.com/user-attachments/assets/031f407b-f7e7-4dfe-a594-c6daef1538ca)

**Limited**
![2025-03-01 18_02_44-Window](https://github.com/user-attachments/assets/cc415bca-8444-4c2d-a56a-d9b7fb68a59b)

**Batch Imports**
![2025-03-01 18_02_25-Window](https://github.com/user-attachments/assets/a276d50c-a751-42d6-b42c-488f84c451d2)
![2025-03-01 17_48_09-Window](https://github.com/user-attachments/assets/210829e4-e5b5-41b2-a2c5-742e2b9d8303)

**Newly added Buttons**
![2025-03-01 17_47_53-Window](https://github.com/user-attachments/assets/13c27c0f-ba18-46df-9385-b767da0fe8f2)

rdiaz738 committed:
Updated to cocurrent downloads attempt 1

rdiaz738 committed:
Update ytdl.py

rdiaz738 committed:
update class

rdiaz738 committed:
Concurrent downloads

rdiaz738 committed:
Added concurrent and limited modes

rdiaz738 committed:
Update main.py

rdiaz738 committed:
Merge remote-tracking branch 'upstream/master'

rdiaz738 committed:
Update app.component.html

rdiaz738 committed:
Update app.component.html

rdiaz738 committed:
Updated ui and backend

Added Sequential, limited and concurrent downloading and import export buttons

rdiaz738 committed:
Delete docker-compose.yml

rdiaz738 committed:
Update README.md

alexta69 commented on README.md:
```diff
@@ -61,8 +61,12 @@ Certain values can be set via environment variables, using the `-e` parameter on
 * __YTDL_OPTIONS_FILE__: A path to a JSON file that will be loaded and used for populating `YTDL_OPTIONS` above. Please note that if both `YTDL_OPTIONS_FILE` and `YTDL_OPTIONS` are specified, the options in `YTDL_OPTIONS` take precedence.
 * __ROBOTS_TXT__: A path to a `robots.txt` file mounted in the container
-
-The following example value for `YTDL_OPTIONS` embeds English subtitles and chapter markers (for videos that have them), and also changes the permissions on the downloaded video and sets the file modification timestamp to the date of when it was downloaded:
```
Was this paragraph dropped by mistake?

alexta69 commented on README.md:
```diff
@@ -61,8 +61,12 @@ Certain values can be set via environment variables, using the `-e` parameter on
 * __ROBOTS_TXT__: A path to a `robots.txt` file mounted in the container
-
-The following example value for `YTDL_OPTIONS` embeds English subtitles and chapter markers (for videos that have them), and also changes the permissions on the downloaded video and sets the file modification timestamp to the date of when it was downloaded:
+* __DOWNLOAD_MODE__ :This flag controls how downloads are scheduled and executed. There are three possible values:
```
Can you please document what the defaults are, for this flag and the next? You can use the same style as is used for the other configuration options: Defaults to <value>.

alexta69 commented on app/main.py:
```diff
@@ -152,32 +162,37 @@ async def delete(request):
-    for _ ,v in dqueue.done.saved_items():
+    for _, v in dqueue.done.saved_items():
         history['done'].append(v)
     for _ ,v in dqueue.pending.saved_items():
```
I see that the same whitespace mishap exists here as above, perhaps worth fixing too while you're at it

alexta69 commented on app/ytdl.py:
```diff
@@ -305,20 +370,20 @@ async def __add_entry(self, entry, quality, format, folder, custom_name_prefix,
+                    if self.config.DOWNLOAD_MODE == 'sequential':
+                        asyncio.create_task(self.__start_download(download))
+                    else:
+                        asyncio.create_task(self.__start_download(download))
```
Looks like both the `if` and the `else` run the same code?

alexta69 commented:
Thanks @rdiaz738 ! It's been a long-requested feature, it's great that you've implemented it.
Maybe it's worth trying to arrange the new buttons somehow into the grid that already exists under the URL bar? Right now it looks like they're detached in their own space and there's some waste of screen real estate. Maybe there needs to be a "batch" drop bar or something like that.

rdiaz738 committed:
Refactored with feedback

rdiaz738 commented:
@alexta69 I fixed all the issues you pointed out and updated the batch button to be next to the add button. 

![2025-03-04 16_15_36-Window](https://github.com/user-attachments/assets/a2634468-b1ac-4832-85b9-f7acad351129)
![2025-03-04 16_15_23-Window](https://github.com/user-attachments/assets/2a4e4845-9463-496c-b394-0e577bc747b0)

alexta69 committed:
Merge branch 'master' into master

alexta69 commented:
@rdiaz738 , this morning #605 was opened after this change, I think perhaps we need to give more space to the buttons (and less to the drop-downs), can you check it?

rdiaz738 commented:
> @rdiaz738 , this morning #605 was opened after this change, I think perhaps we need to give more space to the buttons (and less to the drop-downs), can you check it?
Ya I'm on it, also working on a fix for #607 as well

Skyslycer commented:
Any reason why this doesn't store all URLs immediately on the server and instead slowly queues the URLs? I'm currently downloading roughly 300 videos at once (yeah I know, a bit much), but I thought I could just click "Import URLs" and it would ingest all URLs at once and store them server side, so I don't have to wait for queuing and can close the browser.

rdiaz738 commented:
> Any reason why this doesn't store all URLs immediately on the server and instead slowly queues the URLs? I'm currently downloading roughly 300 videos at once (yeah I know, a bit much), but I thought I could just click "Import URLs" and it would ingest all URLs at once and store them server side, so I don't have to wait for queuing and can close the browser.

Pull request is up for this https://github.com/alexta69/metube/pull/963
